data_diff/databases/clickhouse.py

```diff
@@ -0,0 +1,153 @@
+        if not len(self.cursors):
+            _ = super().cursor()
+        return self.cursors[0]
+
+
+class Clickhouse(ThreadedDatabase):
+    TYPE_CLASSES = {
+        'Int8': Integer,
+        'Int16': Integer,
+        'Int32': Integer,
+        'Int64': Integer,
+        'Int128': Integer,
+        'Int256': Integer,
+
+        'UInt8': Integer,
+        'UInt16': Integer,
+        'UInt32': Integer,
+        'UInt64': Integer,
+        'UInt128': Integer,
+        'UInt256': Integer,
+
+        'Float32': Float,
+        'Float64': Float,
+
+        'Decimal': Decimal,
+
+        'UUID': Native_UUID,
+
+        'String': Text,
+        'FixedString': Text,
+
+        'DateTime': Timestamp,
+        'DateTime64': Timestamp,
+
+    }
+    ROUNDS_ON_PREC_LOSS = True
+
+    def __init__(self, *, thread_count: int, **kw):
+        super().__init__(thread_count=thread_count)
+
+        self._args = kw
+        # In Clickhouse database and schema are the same
+        self.default_schema = kw['database']
+
+    def create_connection(self):
+        clickhouse = import_clickhouse()
+        try:
+            connection = SingleConnection(**self._args)
+            return connection
+        except clickhouse.OperationError as e:
+            raise ConnectError(*e.args) from e
+
+    def _parse_type_repr(self, type_repr: str) -> Optional[Type[ColType]]:
+        nullable_prefix = 'Nullable'
+        if type_repr.lower().startswith(nullable_prefix.lower()):
+            type_repr = type_repr[len(nullable_prefix):].lstrip('(').rstrip(')')
+
+        if type_repr.startswith('Decimal'):
+            type_repr = 'Decimal'
+        elif type_repr.startswith('FixedString'):
+            type_repr = 'FixedString'
+        elif type_repr.startswith('DateTime64'):
+            type_repr = 'DateTime64'
+
+        return self.TYPE_CLASSES.get(type_repr)
+
+    def quote(self, s: str) -> str:
+        return f'"{s}"'
+
+    def md5_to_int(self, s: str) -> str:
+        substr_idx = 1 + MD5_HEXDIGITS - CHECKSUM_HEXDIGITS
+        return f'reinterpretAsUInt128(reverse(unhex(lowerUTF8(substr(hex(MD5({s})), {substr_idx})))))'
+
+    def to_string(self, s: str) -> str:
+        return f"toString({s})"
+
+    def normalize_timestamp(self, value: str, coltype: TemporalType) -> str:
+        if coltype.rounds:
+            prec = coltype.precision
+            timestamp = f'toDateTime64(round(toUnixTimestamp64Micro(toDateTime64({value}, 6)) / 1000000, {prec}), 6)'
+            return self.to_string(timestamp)
+        else:
+            fractional = f'toUnixTimestamp64Micro(toDateTime64({value}, 6)) % 1000000'
+            return f"formatDateTime({value}, '%Y-%m-%d %H:%M:%S') || '.' || {self.to_string(fractional)}"
+
+    def _convert_db_precision_to_digits(self, p: int) -> int:
+        # Done the same as for PostgreSQL but need to rewrite in another way
+        # because it does not help for float with a big integer part.
```

Comment:
I don't understand this comment. Can you explain it? What floats, for example, it doesn't help with?

Reply:
It's dirty hack to save a compatibility with Postgres. 
Clickhouse `float32` use `float` of C language under the hood. I guess C float is used for Postgres' 4-byte `real` (or `float4` that is the same) too because it has the same behavior. It still remains a mystery to me why 2 is [subtracted](https://github.com/datafold/data-diff/blob/master/data_diff/databases/postgresql.py#L49) here for PostgreSQL. I think it's not wierd precision issues as mentioned in a comment for postgres but a normal behavior of 4-byte floating number which has 6 precision. For float it means it can store no more than 6-7 digits in average. Let me clarify on example for Postgres (the same fot Clickhouse):
```
with tmp as (
    select 0.123456789::float4 x
    union
    select 10.123456789::float4 x
    union
    select 100.123456789::float4 x
    union
    select 1000.123456789::float4 x
)
select
    x::decimal(38, 6)
from tmp
order by x;
```

Result:
```
      x      
-------------
    0.123457
   10.123500
  100.123000
 1000.120000
(4 rows)
```
If we increase an integer part 10 times, we leave 1 digit of a floating part. So, 6-7 precision for float means that it can store 6-7 digits for whole number, namely, integer + float parts. 

I see in unit tests that  we have a number [100.98](https://github.com/datafold/data-diff/blob/master/tests/test_database_types.py#L327), and I think tests are passed because of substracting of 2 from precision for Postgres. But if you add a number with a higher integer part, for example, 1000.98, tests failed.

My opinion, we cannot cast `float` to `decimal(38, some_precision)` because precision here depends on the integer part of a float value.

Reply:
It does look like the precision of floats is a little more complicated than we currently account for.

But following your example, I can think of another problem:

```sql
with tmp as (
    select 123456789::float4 x                                                                                                                                                     )
select
    x::decimal(38, 6)
from tmp;

        x
------------------
 123457000.000000
```

So even the integer parts aren't safe!

I do think we can make two assumptions that helps us a lot -

1) The value already sits in some database, and probably was given a datatype that makes sense on both ends

2) We only need to be as precise at the tool that copied it (probably in a migration).

But yeah, I agree this is a weak point, and it deserves some more attention.

Reply:
@pik94 It's not "C language `float`s," it's the IEEE 754 floating-point standard that defines what `float` should look like and how it should behave for all digital computers in order to ensure interoperability.  C (and virtually every other programming language) does it that way because the CPU does it that way.

Reply:
@masonwheeler, yes, it's more accurate to say that it's behavior of the floating point number described in the IEEE 754 standard. I just used `float` because it's mentioned in clickhouse [docs](https://clickhouse.com/docs/en/sql-reference/data-types/float/).